### ISSUE TITLE
[RFC] Stop recommending and using XLIFF for translations

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#22566]
[XLIFF](https://en.wikipedia.org/wiki/XLIFF) is a XML-based format to standardize the localization of software applications. It's allegedly used by professional translation services and it's been recommended by Symfony since the beginning.

### My proposal

1. Stop using XLIFF in our own translation files (validators.xlf, security.xlf, etc.)
2. Stop recommending XLIFF and start recommending YAML instead
3. Don't deprecate or change anything about XLIFF translation (people using it won't notice any change)

### My reasoning

1. We don't use or need any XLIFF feature.
2. We're not fully XLIFF-compliant, much less with the new XLIFF 2 standard (see #15104, #20076, etc.)
3. XLIFF verbosity is unbearable.
4. I guess +80% of Symfony apps don't outsource their translation to external professional services (where using XLIFF could make sense).

Thoughts?

**Comment by netsuo** [comment]
I'm +1 for this. After using XLIFF for a long time, I switched to YAML. A lot more readable, wasn't using any XLIFF-added features.
A lot of translation services supports many formats, including YAML.

**Comment by mvrhov** [comment]
* XLIFF is translation standard, yaml is not. 
* Even Apple switched to xliff. And there are at least two open source tools (weblate and pootle) that you can set up yourself that support it but neither of them will work with proprietary yaml format.
* We should fix the way xliff is handled (JMSTranslationBundle does it correctly)
* xliff allows you to have id separate from the "default string" It's really nice to see the default translation instead of some arbitrary string e.g `dotted.notation.` 
* Putting the english string as a translation when having dotted notation looses the default string

**Comment by javiereguiluz** [comment]
@mvrhov thanks for your comment! We need contrary opinions to make better decisions.

Just a reminder: this proposal wouldn't change or deprecate or remove XLIFF support. You could keep using it exactly as before. Nothing would change for you. This proposal is just to stop using it ourselves for Symfony translation files (we don't use or need any XLIFF feature) and to stop recommending it by default.

**Comment by ScullWM** [comment]
@javiereguiluz Even if using external professional services, most of them support YAML format.

**Comment by theofidry** [comment]
👍 to recommend YAML by default. But like for the services declaration, I would keep it clear that XLIFF & XML are the way to go for libraries/bundles.

**Comment by dunglas** [comment]
👎, we're committed to implement, support and promote open standard. XLIFF is an industry leading and open standard for translation. Until a better one emerge, we should keep it as default.

**Comment by iltar** [comment]
I don't think people will be using XLIFF tools to modify the translations in Symfony repositories, they do it manually, probably in a text editor of sorts. I don't see an added benefit to keep using XLIFF in Symfony itself for the translations, while I do see a lot of benefits of switching to yaml. 

* Easier to modify, you don't have to know the structure
* Less chance of syntactic typos
* Easier to create dumps
* Easier to parse, export and import.

As mentioned, XLIFF won't be removed nor deprecated, it's just about what is used in Symfony itself. I'm all :+1: for that.

**Comment by Simperfit** [comment]
👎 on this one, XLIFF is the standards in translation, why should be switch to something that is not build to do that. 

There's a lot of open source tools to modify XLIFF.

**Comment by javiereguiluz** [comment]
@dunglas you said *"XLIFF is the industry leading open standard for translation"*. I'd like to have more info about that. Do you have some link with stats or a list of the main open source projects using XLIFF? Thanks!

**Comment by iltar** [comment]
> There's a lot of open source tools to modify XLIFF.

But how many people actually use XLIFF tools to manage the translations in the Symfony repository? Or perhaps to even read them? Reading XML is really annoying and I'm not going to use a tool just for the translations to read them.

**Comment by stof** [comment]
> But like for the services declaration, I would keep it clear that XLIFF & XML are the way to go for libraries/bundles.

@theofidry Most bundles I know that are providing translations are actually already using YAML rather than XLIFF (including my own bundles) because they don't rely on professional tools to maintain translations and editing by hand is far easier for a YAML file.

**Comment by mickaelandrieu** [comment]
> Just a reminder: this proposal wouldn't change or deprecate or remove XLIFF support. You could keep using it exactly as before. Nothing would change for XLIFF users. 

I'm not sure of it. If it's not recommended and not used anymore, the next step will be to deprecate it arguing "this is not used anymore". I hope Symfony Translation component will or should support Xliff 2: we had to do so much in the PrestaShop project cause the current support is limited to basics and hard to extend.

Then you said - and I disagree with you - 'We don't use or need any XLIFF feature': theses features aren't officialy implemented, but I can tell that at least three big french companies use their own xliff dumpers and builders to use specifics properties of Xliff for translations. Maybe this means Xliff format should have his own component (like Yaml), I don't know.

**Comment by javiereguiluz** [comment]
> the next step will be to deprecate it arguing "this is not used anymore"

We have a ton of loaders for all kinds of translation formats. We don't use them, but we won't deprecate them: https://github.com/symfony/symfony/tree/master/src/Symfony/Component/Translation/Loader  Same for XLIFF.

> We don't use or need any XLIFF feature

I meant Symfony. The Symfony project in this GitHub repository has tens of XLIFF translations files. But we don't need or use any XLIFF feature. Of course that companies and developers using XLIFF use and need its features.

**Comment by stof** [comment]
@mickaelandrieu what is used inside Symfony itself does not determine what is used in the whole ecosystem. We have several loaders which are not used y Symfony itself and are not the official recommendation. This does not mean that they should get deprecated.

> I can tell that at least three big french companies use their own xliff dumpers and builders to use specifics properties of Xliff for translations.

If they already have working implementations supporting some XLIFF features, do you think they could open PRs to contribute them to Symfony itself ?

**Comment by ro0NL** [comment]
:+1: 

Except for

> Stop recommending XLIFF and start recommending YAML instead

Why not recommend both? Explaining when to use one or another (basicaly simple vs. advanced usage).

**Comment by dunglas** [comment]
@javiereguiluz it's an OASIS standard (a well known organisation).

* Angular (2+) uses it by default (https://angular.io/docs/ts/latest/cookbook/i18n.html)
* Apple recommends XLIFF to localize Mac OS and iOS apps (https://developer.apple.com/library/content/documentation/MacOSX/Conceptual/BPInternational/LocalizingYourApp/LocalizingYourApp.html)
* Google recommends XLIFF for Android apps (https://developer.android.com/distribute/marketing-tools/localization-checklist.html)

We may switch to a another standard alternative (TMX or something like that), but using and promoting a custom format would be a step backward.

**Comment by Simperfit** [comment]
@dunglas @javiereguiluz [TYPO3](https://typo3.org/) is using XLIFF too

**Comment by ro0NL** [comment]
Who's using XLIFF is not the point. It's about what SF is going to use itself...

**Comment by Simperfit** [comment]
@ro0NL Yup, but @javiereguiluz was asking who's using it :).

**Comment by javiereguiluz** [comment]
@Simperfit I asked that because @dunglas said *"XLIFF is the industry leading open standard for translation"*. I wanted some stats to check if that was true. I'm still looking for some stats about the translation format used by open source projects.

**Comment by JuliusKoronci** [comment]
XLIFF is supposedly better but now one is using it anyway..and xml-ish style is just not fit for the 21st century

**Comment by Neirda24** [comment]
I agree that XLIFF is wayyyy to verbose and YAML can provide very good features. However I'm the `ISO` kind of guy... I love that symfony respects so much standards. It make it easier for developper team to bond and agree on `howto's`. I personnaly never outsourced the translations so I guess I never needed it fully (the XLIFF) but if it is the standard then I think we should use it. so 👎  for me.

**Comment by ogizanagi** [comment]
It seems everyone is going crazy because they kind of think this suggestion will make the XLIFF format "obsolete" or not maintained as well as before anymore. But this is just about avoiding to promote blindly a format  that is way too verbose for most use-cases, and for which promoted features aren't actually used by most bundles maintainers nor Symfony itself.

Let's just mention each format advantages, without recommending anything, and let developpers use whatever is the more suitable for their needs. For my own experience, I never ever used the XLIFF format nor professional tools for translations, nor my clients did.

**Comment by mvrhov** [comment]
@ogizanagi. You know in our country we have a saying.  Direct translation would be "The donkey will go on the ice only once". I believe that the English one is "Fool me once, shame on you; fool me twice, shame on me".  So when FQCN == service id was introduced I was assured that nobody would force me to use that. 2 months later, there is a bunch of new features with autowiring that forces me to do exactly that, or write an alias for shor service id which equals FQCN for each one, if I want to use autowiring.

**Comment by rodrigobb** [comment]
I'm going to change from upvote to down vote. My previous reason was that I'm using yml, so laziness. But people against this proposal have convinced me. Their reasons are much better and Symfony should remain close to standards.

**Comment by javiereguiluz** [comment]
@rodrigobb I'm curious about something: are you voting for XLIFF *and* changing your YAML files to XLIFF ... or are you going to keep using YAML but recommending others to use XLIFF?

**Comment by Nyholm** [comment]
I read two discussion here. 

## 1) Should Symfony translation files use Yaml or XLIFF?

* We currently have XLIFF. 
* XLIFF is the standard for translations. 
* Yaml does not give any technical benefits. 

Sure Yaml might be easier to read, but who reads the Symfony translation files anyways?
I vote for keeping XLIFF but Im fine either way. 

## 2) Should we recommend using Yaml or XLIFF?

For small apps with one or two languages, sure Yaml is the way to go. You will edit these files and regularly work with them. No question about it. 

But, when you are doing something more than a hobby project you need to take translation more seriously. You probably do not keep translations in Git because you are using a translation service. You are probably aslo using extractors and a [WebUI](http://php-translation.readthedocs.io/en/latest/symfony/webui.html) to add new translations to the translation service. So for a real project you want to use XLIFF because your translation tools can store extra data in the XML document. At this point you do also not care what format the translation files are because you do never work with them directly. They are just a local database. 

Lets recommend Yaml for small apps and then recommend a few options for something cool and useful when the app is larger. We at [PHP-translation](http://php-translation.readthedocs.io/en/latest/index.html) is trying to be one of those options.

**Comment by fabpot** [comment]
@Nyholm Exactly what I think!

**Comment by jrobeson** [comment]
Can't yaml format be converted losslessly to xliff?  Although the reverse is not true.

**Comment by hhamon** [comment]
I'm really not convinced by YAML for translation files especially when it comes to include complex translated contents.

XLIFF is a standard and it's XML based, which means:

1. it's easy to parse and validate natively with PHP,
2. It allows to include meta information about a translation (comments, notes, etc.). YAML doesn't support meta information.
3. It's extensible.
4. It's portable and easy to edit in GUI tools.

So yes it's verbose compared to YAML but it's still a standard format. We, the Symfony community, must continue to promote standard tools.

**Comment by iltar** [comment]
If I read most of the comments here, it seems like it forgets about the most important point to me:

> 1. Stop using XLIFF in our own translation files (validators.xlf, security.xlf, etc.)

Let's take a new bundle in Symfony which has some translations, which I want to add  new translation file to. I'm merely a developer that often works in vim and PhpStorm and I don't use fancy tools, but I want to contribute to symfony so that my own language is supported (scenario).

If I were to have the XLF file and edit it as plain text file, I can make a lot of mistakes and I have no clue what some of the things are. When translating, I can do a lot of copy-paste, but when creating a new file, I have no idea what to do.

I will take the [validators.en.xlf](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Resources/translations/validators.en.xlf) as base file.

```xlf
<?xml version="1.0"?>
<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
    <file source-language="en" datatype="plaintext" original="file.ext">
        <body>
            <trans-unit id="28">
                <source>My custom source translation.</source>
                <target>My custom target translation.</target>
            </trans-unit>
        </body>
    </file>
</xliff>
```

A few things in this file:
 - `xliff version="1.2"`, no idea what I need for this, I'll just leave it.
 - `datatype="plaintext"`, I'll just leave this, I guess this is right?
 - `original="file.ext"` no idea what this is for, I'll just leave it.
 - `trans-unit id="28"` what kind of id? I guess I'll just wait for someone to review it and say it's wrong and what I should do with it.
 - High chance of making a typo in one of the tags or attributes.

Now, if I were to have this file in yaml, it would be _a lot_ easier to contribute my translations to Symfony:

```yml
"My custom source translation.": "My custom target translation."
```

### Conclusion
I don't really mind what Symfony recommends for developers, I feel like Symfony should promote the possibilities and their pros and cons, let the developer decide. Recommend XLIFF to new developers and they will pull their hair out. Recommend Yaml to a company that works with professional tools, and they might look for another framework.

However, for Symfony, the developers are not professional translators. We are contributors and developers, we love writing in tools like vim or notepad and our precious IDE, which we find superior to whatever another developer is using. For Symfony it should imho be very easy to contribute to translations if needed. With Yaml I believe that contributing is easier and it won't influence anybody who is using professional translation tools. If they want to customize translations from Symfony, they'll have to overwrite them with their own file anyway, ignoring the Symfony variant.

We don't use any fancy XLIFF functionality in Symfony, it's pretty much still a key-value replacement, even with the counts.

**Comment by robfrawley** [comment]
I have never used XLIFF, including on larger projects (even one that was for a company that created a Symfony-based publishing workflow) for one important reason: it's a pain to work with. YAML is simple and anyone, whether they were a new employee with less than a day on the job or a veteran with 10+ years, can easily make changes without issue.

:+1: For Symfony using YAML internally and promoting its use. That said, I'd suggest Symfony maintain documentation on basic usage of XLIFF.

**Comment by mvrhov** [comment]
@iltar: That just means that you are being ignorant. You don't start coding without reading some documentation on the language. or you just don't start using the Symfony without reading some docs first.

**Comment by Pierstoval** [comment]
@ogizanagi said:
> For my own experience, I never ever used the XLIFF format nor professional tools for translations, nor my clients did.

And I think it's true for most companies. In my previous jobs where I developed tons of SF apps, I have **never** used XLIFF format, always Yaml, and it was fine.

Except once: the client herself wanted to pay a translator, and we provided an easy access to `.po` and `.mo` translation formats, and we just exchanged occasionnally some e-mails to warn her about new translations, she updated it, we incorporated them in the app, it was really straightforward.

The point is: it was a **specific case**. So for 99% of all our cases at that time, Yaml was enough.

@ro0NL said:
> Why not recommend both? Explaining when to use one or another (basically simple vs. advanced usage).

I'm totally 👍 to recommend Yaml for small projects, and Xliff for standards (and to avoid systematic Yaml dependency).


@mvrhov said:
> You don't start coding without reading some documentation on the language. or you just don't start using the Symfony without reading some docs first.

I don't know if you're just a developer, lead tech or anything, but I wish you could meet tons of students learning Symfony or small & medium-class companies that use Symfony, and tell them _"Don't stard coding without reading docs"_.
Lots of students will tell you that docs are for stupids, being arrogants (even though it's not the majority), but most of them simply don't comply with the idea reading one another pile of docs when they already have lots of books & courses to read, especially when they can just rely on a simple and readable format like Yaml.
And for the case of companies, just tell them this sentence and they will yell at you that customers have delay, that they just want a _"Plug-and-play"_ solution, or that _"As long as it works at first sight, don't mind me to encourage spending more time for a more complex stuff that brings no money"_.

Even if I agree that XLIFF is a professional and standard format, Yaml is much better for DX. And what are we? Developers! What do we want? Code performant apps fast! And for this, we need simplicity. Yaml is simplicity. So even though I don't agree that Yaml should be the "first recommended format", I completely understand people saying that it should.

> P.S. Having default translations in Yaml you are making it difficult for those who use the professional tools to do translations. The Translation memory and other features really make it worthwhile

This is not an argument: when you start coding 99% times you will **already know** whether it's gonna be translated, and **how** it will be. At that time, if you already know that your client will send you an Excel file, just put translations in CSVs. Else, if you know you'll have to maintain translation with an external company, use **their** tool. Else, use the one that is the **simplest** and **fastest** to use: Yaml.

---

That being said, completely agree with @iltar's proposal: don't recommend anything, just propose the two most famous formats Yaml and XLIFF as the most used, explain the use-cases in one line (Yaml for DX & small projects, XLIFF for standards & professional tools) and that would be clearly enough for the community and companies that will use Symfony.

### My conclusion

Propose the two formats and explain pros and cons is the best option to me. And then, tell that we have lots of other formats.

**Comment by Pierstoval** [comment]
Sorry four double-posting but it's to avoid editing a maybe already-read long post:

@mvrhov:
Just to say and add my 2cents, ask for example @afilina how many times she rescued ugly & 💩 projects, and ask her about relationships between coders and marketing directors or project managers. Most of the time, companies _don't have time_, and reading docs needs time.
There are companies where developers just code because it's something they kinda _like_ and they kinda _know_, and when they use something like Symfony, they just use it, read a _"Get started"_ guide, start coding, deploy 💩, but don't care, because "it works" and that's all.
In this discussion, we're representing a **minority**: the passionated ones that want things to be clear, clean, concise, simple, standard, performant, sexy, whatever. The **majority**, however, is silent and won't even care about what we're talking about right now, they won't even care about new Symfony versions. They use Symfony because "it's popular & standard", code, ship, it works, _basta_, and projects managers are happy and clients are happy.
If you encounter _this_ majority, I bet you will be upset about the real world of web agencies & IT companies, but trust me: simplicity is the best option.

So in this case, as contributors for a **standard** tool, we **should** recommend standard tools. But also, as contributors for a **developer** tool, we **should** also recommend DX-friendly tools.

That's why I re-post my conclusion a second time:

Recommending the two formats, Yaml and XLIFF, and explain "how to choose" in a one-liner concise sentence is the best option to satisfy 99% developers.

**Comment by rodrigobb** [comment]
@javiereguiluz I've been working in projects using XLIFF and Yaml formats and I've adopted both formats for my own code (when I was developing a new bundle/module and, of course, extending the current file where in current bundles).

When I work alone in a project I usually choose Yaml because is faster and closer to the developer (me). And because no external people will have to work with that files. 

However, if I start a new project and I know it will be a complex project (related to translations complexity, I mean) or I suspect translations will lie over third people (mainly non technical people), I will use XLIFF without any doubt.

**Comment by adamelso** [comment]
Having read both sides of the discussion, I'd support an RFC along the lines of what @Nyholm mentioned. However the details outlined in this RFC feel like we're crossing that fine line where the "exception becomes the rule", so I'm 👎.

Someone mentioned creating a console command that converts XLIFF to YAML and vice versa - sounds nice if feasable to have one.

**Comment by javiereguiluz** [comment]
I'm closing this proposal as "won't fix" because there is not enough consensus. A big change like this needs a 5-to-1 or 10-to-1 consensus and we are only 2-to-1.

Thank you all for the discussion.

**Comment by cdeyoung** [comment]
I know this is closed, but I wholeheartedly agree with the shift to YAML.  XLIFF (and all other uses of XML) is a bloated waste of time, space, and energy.  YAML and JSON are both better formats (in that order, in my opinion), especially for localization purposes.  XML is an abuse-prone solution to a problem nobody ever had.

**Comment by Pierstoval** [comment]
> XML is an abuse-prone solution to a problem nobody ever had.

This is probably because you don't _actually_ know or use XML.

Yaml and JSON have no benefit other than "readability". XML, even as verbose as it is, IS readable, and more: it has a nice standard plugged to it with XSD that make the definition of all tags & attributes documented and compatible with ALL tools that have been in here for 20 years or less (as XML started in 1996).

With JSON we could at least expect the use of a json-schema, but it's not widely used, and Yaml has no self-validation whatsoever.

XML, and particularly XLIFF, have a really big advantage.

It's just that for translations you _may_ not need it as we _may_ just be okay with just key=>value pairs. 

But the issue was closed because not everyone agrees on this, and I think it's ok to keep it as-is:

* XLIFF for open-source projects, to avoid having Yaml as a dependency
* Yaml for projects with no external translation management (and it's the default anyway)
* The `translation:update --output-format=xlf|yaml` command to convert from a format to another

I think this will suit to 90%+ people 🙂

**Comment by msoutopico** [comment]
> 
> 
> @dunglas you said _"XLIFF is the industry leading open standard for translation"_. I'd like to have more info about that. Do you have some link with stats or a list of the main open source projects using XLIFF? Thanks!

This might be of help, from the point of view of translators: http://www.internationalwriters.com/toolkit/10_October_Standards.pdf

**Comment by msoutopico** [comment]
Have a look at OmegaT. It's a open-source translation editor that leverages translations memories (TM) and has filters for many formats, including XLIFF and YAML (other similar tools might have those two too). That means your files could be translated in this tool (using TMs and all other translation functions) regardless of what format you use to share your files. XLIFF is a good choice because any translation tool is meant to understand it, but it is an intermediary format meant to separate localizable content from structure and to share the translation file with any third party (translator, reviewer, LQA, etc.). However, if the translation tool has a filter for the original source (YAML, JSON, etc.), the intermediary format is --still useful, but-- less necessary. Another advantage of XLIFF (as well as TMX) is that it is bilingual, so it's not necessary to use a pair of files (source and target).

**Comment by comradekingu** [comment]
@mvrhov As per your comment above, YAML and "Ruby YAML files" are now supported in Weblate https://docs.weblate.org/en/latest/formats.html?highlight=yaml#translation-types-capabilities

@msoutopico I have used OmegaT professionally, but I like Weblate better as something to attract more translators with an easier workflow. The point remains to use or recommend _something_.

https://symfony.com/doc/current/translation.html and https://symfony.com/components/Translation leaves a bit to be desired.

**Comment by 537mfb** [comment]
Sorry to resurrect this thread, but this is a subject that should be revisited every once in a while - and with version 6 of Symfony already about to be released, I think this a good opportunity for this.

I actually work at a translation agency and our CAT tool can work with xliff just fine, but also works with json, yaml and php - so any of those formats would be fine to receive from a client, and I don't think this is anything specific about the CAT tool we use - any CAT tool worth anything this days should be able to handle at least json and yaml on top of xliff if not php also.

What we have a bit more trouble is when a client sends us files but then says he only wants to translate some parts of it and not others. This takes us a bit more time to prepare a quote, but fortunately our CAT tool allows us to create Regex based rules to handle such cases - it can be a bit of work to find exactly what rules should be used and needs to be done on a client by client base but we've done it several times.

Having said that, the excuse of recommending using xliff because of external translators should really be discarded - if your translators can't handle json and yaml they must either not be using a CAT tool or using a version of one that's at least closing in onto 10 years old by now, and neither of this is a good sign.

Also, using json, yaml or php would make the recommendation of using keys actually make sense. They could be simply imported into a CAT tool for translation, while an xliff with keys would just create difficulty to your translator.

Where I work, if the xliff has the actual text in source we can just import it, and the tool gets the text to translate from the source and places the translation in the target, but if the source is keys, I have to create an import rule to get the text from target instead of source and still place the translation in the target, which can have problems if the target has specially encoded characters and I don't notice. I usually apply an HTML tagger filter on top of my custom filter to handle this types of situations but I still have to go through it to make sure everything is ok, and all of this takes time that might add to our quote.

**Comment by 537mfb** [comment]
The recommendation should be:

If you're using keys go with yaml, otherwise go with xliff
Using dotted notation keys makes more sense with yaml (or php, or json), while using the default text makes more sense in an xliff

**Comment by Nyholm** [comment]
Thank you. 
I understand your point of view. For Symfony 6 specifically it is a bit too late Im afraid. But topics like this should be revisited time to time. 

XLIFF is [designed to handle translations](https://en.wikipedia.org/wiki/XLIFF). That cannot be said for json or yaml or PHP files. I do understand that some other formats are easier to work with in some situations. 

Note that Symfony is not require you to use XLIFF, it is just a recommendation. Feel free to continue the discussion but please open a new issue.

**Comment by 537mfb** [comment]
Hi @Nyholm

XLIFF was designed to handle translations - true - but that implies the source is the original text.

When using the dotted notation keys then XLIFF is not designed for that, and if used with keys instead of the original text in the source the point of using it for external translation tools/professionals becomes mute, as at that point those professionals can't treat it as an XLIFF anymore

That was my point - if using the dotted notation keys then something like YAML or JSON is a better fit

**Comment by welcoMattic** [comment]
Hi @537mfb,

We are using XLIFF with dot notation on all Translation Providers, it works pretty well. All SaaS (Loco, Lokalise, Crowdin for now) are able to treat XLIFF with dot noted source.

I think we should continue to recommend (but not require) using XLIFF, as it's a standard for translation file. But feel free to use json or yaml if it suit better for your needs.

**Comment by 537mfb** [comment]
Hi @welcoMattic 

I never said dot notation didn't work in XLIFF (I've used it in the past) - I said that's not what XLIFF's where designed for

I was answering @Nyholm where he was pointing out that XLIFF's where designed for translation - and I pointed out that when using dot noted source, that's not what XLIFF's were designed to be anymore.

At that point technically it's not really an XLIFF anymore, just an XML that follows the structure of an XLIFF but has lost all it's value as an exchange format for professional translation tools - It's just adding the verbosity of xml while removing the value that an XLIFF is supposed to bring to the table.

I work at a translation agency - and that's where I'm coming from here.

When using dot noted source, a json or yaml would be a matter of simply importing to the CAT tool and start working. An XLIFF with dot noted source is a mess requering thinkering with regex to create a custom filter for the CAT tool - doable sure, but adds extra work even before there is a quote, and that work needs to be invoiced.

When using the default text as source, then sure, that's the proper use of XLIFF and if it ever goes to a professional translator they'll be thankfull you used XLIFF

If using the dotted notation source, YAML or JSON is a better fit both for editing by a developer and to send to a professional translator - outherwise XLIFF is the better choice on both counts (IMHO keys on YAML or JSON should never be sentences)

As I said, the recommendation should be:

If you're using keys go with yaml, otherwise go with xliff
Using dotted notation keys makes more sense with yaml (or php, or json), while using the default text makes more sense in an xliff

If those files ever need to be sent to an external professional translator, they'll appreciate it
